tokio/src/sync/batch_semaphore.rs

```diff
@@ -188,58 +185,61 @@ impl Semaphore {
 
     /// Release `rem` permits to the semaphore's wait list, starting from the
     /// end of the queue.
-    ///
-    /// This returns a new `LinkedList` containing all the waiters that received
-    /// enough permits to be notified. Once the lock on the wait list is
-    /// released, this list should be drained and the waiters in it notified.
     ///  
     /// If `rem` exceeds the number of permits needed by the wait list, the
     /// remainder are assigned back to the semaphore.
-    fn add_permits_locked(
-        &self,
-        mut rem: usize,
-        mut waiters: MutexGuard<'_, Waitlist>,
-    ) -> LinkedList<Waiter> {
-        // Starting from the back of the wait queue, assign each waiter as many
-        // permits as it needs until we run out of permits to assign.
-        let mut last = None;
-        for waiter in waiters.queue.iter().rev() {
-            // Was the waiter assigned enough permits to wake it?
-            if !waiter.assign_permits(&mut rem) {
-                break;
+    fn add_permits_locked(&self, mut rem: usize, waiters: MutexGuard<'_, Waitlist>) {
+        let mut wakers = [None, None, None, None, None, None, None, None];
```

Comment:
length of the array chosen totally arbitrarily...if anyone has a number that they think is faster or more aesthetically pleasing, i can change it.

Reply:
:shrug: seems fine.

Reply:
Is there no way to write like `[None; 8]`? I would probably have used a higher number since it's not that much stack (a `Waker` is 16byte). But I doubt it matters in practice. If you have too many wakers woken up at once its probably a bad design anyway :-)

Reply:
Unfortunately, `[$expr; n]` initializers work by evaluating `$expr` once and then cloning it `n` times, so the compiler didn’t like that here. The array was going to have 16 elements but i didn’t wanna type “None” 16 times. :P

Reply:
it works w/ `Default::default()`.

```
let mut wakers: [$ty; 8] = Default::default();
```

Reply:
Oh, that's much nicer, I didn't think of that.